### ISSUE TITLE
Centerline crashes are not being associated to service road polygons

### ISSUE THREAD

**Issue by frankhereford** [cityofaustin/atd-data-tech#6763]
Testing after the resolution of issue #6759 has revealed that crashes which have been geocoded to the centerline of level 5 roads which have a directionality indicated in the road name are not being associated to a location polygon, as would be expected based on this code: https://github.com/cityofaustin/atd-vz-data/blob/master/atd-events/crash_update_location/app.py#L329-L330.

**Comment by frankhereford** [comment]
My initial reaction was that there must be a logical problem in the python script that is responsible for this action, however, the script runs as expected when executed from the command line after being provided the required environment variables. Because of this, I would like to propose that we get on AWS and ensure that the script is running in that environment as we expect and that it's able to run until completion. 

I mention running to completion because this process, as it's implemented now, is a slower process, largely due to the use of a view (`cr3_nonproper_crashes_on_mainlane`) to figure out what polygon to use for a given centerline crash. The use of a view was chosen because it has secondary benefits of being able to be viewed in GIS software and provides the potential for use in SQL JOIN clauses. However, the downside of the view is that many crashes must be inspected and this inspection requires the use of ST_TRANSFORM() into SRID 2277, thereby causing any geometric indexes that may be in place to be unusable. A more performant solution, based on the use of a function to do inspection of a crash, may be required to make the database tasks asked for by the lambda function faster, and we could leave the view around for the secondary benefits mentioned above. Making the lambda function faster will also grant a cost reduction on the AWS bill, since lambda is billed, in part, by the millisecond of execution.

I must emphasize that my ideas here are based purely on speculation, and it will require access to cloudwatch and possibly some debugging statements in the python script to see if they are well founded or not.

**Comment by patrickm02L** [comment]
Talk more about this on backlog grooming on Wednesday, Aug 4th. @patrickm02L

**Comment by sergiogcx** [comment]
@patrickm02L I've made some basic troubleshooting on this code. 

TL;DR:
- AWS is dropping some SQS requests (they are taking too long to execute)
- We can increase the allowed execution time, but that will only create more costs and it will burden the database more as well as not really fixing the problem.
- The database query being used is non-performant, upon close inspection it needs lots of optimizations.

With these conditions, I cannot recommend this code for production. We need to roll back or to refactor the SQL query Frank developed for better performance.

### Errors & Throttles
I noticed a spike in errors at some points of the day, and an equivalent spike in throttles. Errors I don't mind too much, but the throttle worries me because that means the executions are taking too long, which is a direct result of bad database performance. _In pain english, we are burdening the database too much_.

![2021-08-05_12-06-22.png](https://images.zenhubusercontent.com/5b7edad7290aac725aec290c/4973aaac-8219-4553-8ba1-4aacd28ec590)

![2021-08-05_12-07-05.png](https://images.zenhubusercontent.com/5b7edad7290aac725aec290c/c620ac36-0c08-46c6-b94b-1b5b5413f11f)


### Query Performance
Initially, I ran a simple query that would give me 10 rows in order, that took more than 40 seconds:

![2021-08-05_11-00-31.png](https://images.zenhubusercontent.com/5b7edad7290aac725aec290c/41f4a13c-372a-4ea3-a1c3-392a791e175e)


Secondly, I did some digging in the performance metrics and I could identify some bottlenecks, meaning that some parts of the SQL query run in milliseconds, and one or two run very slow:

![2021-08-05_12-03-51.png](https://images.zenhubusercontent.com/5b7edad7290aac725aec290c/b347c0b0-5afc-493d-b225-862ea1b0d444)

![2021-08-05_12-03-06.png](https://images.zenhubusercontent.com/5b7edad7290aac725aec290c/8e3dceaa-f294-4f06-9066-100721761819)

**Comment by sergiogcx** [comment]
@patrickm02L @mateoclarke After discussing these things with @frankhereford, we determined that it was best to refactor the query and re-write it as SQL function with all the optimizations we need.
